### PR TITLE
block: Implement backend for Virtual Machine DisK (VMDK)

### LINKED ISSUE #7167: VMDK support on Cloud Hypervisor
EROFS could take advantage of an image format that supports split files, which is currently not the case for QCOW2.

VMDK specs: https://web.archive.org/web/20240527050638if_/https://www.vmware.com/app/vmdk/?src=vmdk
EROFS context: https://lore.kernel.org/linux-erofs/20250612073103.2538455-2-hsiangkao@linux.alibaba.com/

### PR DESCRIPTION
**Pull request by Sumynwa**
Implement backend for VMDK virtual block format to support FLAT (fixed/pre-allocated sized) disks.
This PR does not implement the complete specification.

In summary, the implementation

1. supports 2 createType: monolithicFlat & twoGbMaxExtentFlat
2. implements synchronous I/O, does not support async_io/io_uring
3. adds unit tests and integration testing

The implementation references online specification from https://web.archive.org/web/20240527050638if_/https://www.vmware.com/app/vmdk/?src=vmdk

Fixes #7167

### CODE REVIEW

File: block/src/formats/vhd/footer.rs [hunk 1]
```diff
@@ -119,6 +119,16 @@ impl VhdFooter {
 
 /// Determine image type through file parsing.
 pub fn is_fixed_vhd(f: &mut File) -> io::Result<bool> {
+    // A fixed VHD stores its footer in the final block of the file, so the file
+    // is always at least one block long. A file smaller than one logical block
+    // cannot be a fixed VHD -- for example a VMDK *descriptor*,
+    // which is a few hundred bytes (well under 512). Bailing out early to
+    // avoid `VhdFooter::new` failing with `EINVAL` for such a short file
+    // aborting image-type detection before the VMDK probe runs.
+    if f.metadata()?.len() < SECTOR_SIZE {
```
Comment (Camelron) [inline block/src/formats/vhd/footer.rs]: Should we be relying on the file size for this check? Is there a better way to do this? I'm imagining all sorts of nightmare bugs where the VMDK happens to be larger than SECTOR_SIZE or in non-VMDK cases with a VHD on a block device which isn't a regular file (--disk path=/dev/sdX).

Maybe we should be probing for VMDK before we reach this branch to avoid this edge case. Or one tweak you could do: (AI suggests this) remove this size check and instead handle it below when we try and make the footer:

```
let footer = match VhdFooter::new(f) {
    Ok(footer) => footer,
    Err(e) if e.kind() == io::ErrorKind::InvalidInput => return Ok(false),
    Err(e) => return Err(e),
};
```

Comment (Sumynwa) [reply block/src/formats/vhd/footer.rs]: Yes this is incorrect check and somehow made it after my cleanup
I will correct this thanks @Camelron

Comment (Sumynwa) [reply block/src/formats/vhd/footer.rs]: Fixed this. Thanks.

File: block/src/formats/vmdk/mod.rs [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+
+// VMDK uses a synchronous, extent-aware worker (see `engine_sync`) that maps
+// each request to the backing extent(s). This supports both single-extent
+// `monolithicFlat` and multi-extent `twoGbMaxExtentFlat` images.
+//
+// io_uring/AIO is disabled for VMDK
+#[derive(Debug)]
+pub struct VmdkDisk {
+    inner: FlatVmdk,
+    use_async_io: bool,
```
Comment (weltling) [inline block/src/formats/vmdk/mod.rs]: This field looks unused.

Thanks

File: block/src/formats/vmdk/mod.rs [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+
+impl disk_file::DiskSize for VmdkDisk {
+    fn logical_size(&self) -> BlockResult<u64> {
+        Ok(self.inner.virtual_block_size())
+    }
+}
+
+impl disk_file::PhysicalSize for VmdkDisk {
+    fn physical_size(&self) -> BlockResult<u64> {
+        Ok(self.inner.physical_block_size())
```
Comment (weltling) [inline block/src/formats/vmdk/mod.rs:48]: `PhysicalSize` is supposed to deliver the true host allocation size, so `query_device_size` should be used here.

Thanks

File: block/src/formats/vmdk/flat.rs [hunk 4]
```diff
@@ -0,0 +1,570 @@
...
+// where it is blocked (`EPERM`, e.g. a seccomp filter) -- it falls back to a
+// per-component `openat` + `O_NOFOLLOW` walk from the same anchor.
+fn open_extent(
+    base_path: &str,
+    filename: &str,
+    writable: bool,
+    direct: bool,
+) -> io::Result<AlignedFile> {
+    let (anchor, resolve) = if Path::new(filename).is_absolute() {
+        ("/", RESOLVE_NO_SYMLINKS)
```
Comment (weltling) [inline block/src/formats/vmdk/flat.rs]: Wouldn't this reject any absolute path that has a symlinked directory in it? That might reject legitimate images. Perhaps only refuse when the final component is a symlink?

Thanks

File: block/src/formats/vmdk/mod.rs [hunk 5]
```diff
@@ -0,0 +1,261 @@
...
+        let dir = TempDir::new_with_prefix("/tmp/vmdk-test").unwrap();
+        let path = write_flat_vmdk(
+            dir.as_path(),
+            "monolithicFlat",
+            &[("disk-flat.vmdk", "RW", 2048)],
+        );
+        let disk = VmdkDisk::new(open_descriptor(&path), &path, false, false).unwrap();
+
+        assert_eq!(disk.logical_size().unwrap(), 2048 * SECTOR);
+        assert_eq!(disk.physical_size().unwrap(), 2048 * SECTOR);
```
Comment (weltling) [inline block/src/formats/vmdk/mod.rs]: This asserts `st_size`, not allocated blocks. Ties to the physical_size/query_device_size comment.

Thanks

File: block/src/formats/vmdk/flat.rs [hunk 6]
```diff
@@ -0,0 +1,570 @@
...
+        )
+    };
+    if ret < 0 {
+        return Err(io::Error::last_os_error());
+    }
+    // SAFETY: `openat2` returned a fresh descriptor we now own exclusively.
+    Ok(unsafe { File::from_raw_fd(ret as RawFd) })
+}
+
+fn open_extent_walk(
```
Comment (weltling) [inline block/src/formats/vmdk/flat.rs:190]: On modern kernel only the `openat2` path runs, so this fallback is untested. It should be either dropped or tested equally.

Thanks

**Review by weltling — CHANGES_REQUESTED** [review]
Nice work. This is in good shape, and the extent path security handling is well done.

Besides the inline comments, the only gap I'd mention - no test for the new zero-sector/overflow rejection in FlatVmdk::new. Worth adding.

Thanks!

File: block/src/formats/vmdk/flat.rs (hunk 4 above)
Comment (Sumynwa) [reply block/src/formats/vmdk/flat.rs]: Makes sense. I will handle that. Specs suggest using only relative path for extents
But implementation like in qemu do allow absolute paths also

File: block/src/formats/vmdk/flat.rs (hunk 6 above)
Comment (Sumynwa) [reply block/src/formats/vmdk/flat.rs:190]: Makes sense. I have unit tests
Will extend to test openat as well.

File: block/src/formats/vmdk/mod.rs (hunk 2 above)
Comment (Sumynwa) [reply block/src/formats/vmdk/mod.rs]: I will remove this.

**Comment by Sumynwa** [comment]
Thanks @weltling i will add tests for them.

File: block/src/formats/vmdk/engine_sync.rs [hunk 7]
```diff
@@ -0,0 +1,357 @@
...
+    }
+
+    fn fsync(&mut self, user_data: Option<u64>) -> AsyncIoResult<()> {
+        // Flush every extent: a single guest flush must durably persist data
+        // that may have been written across multiple extent files.
+        for extent in self.extents.iter() {
+            // Skip NoAccess extents, which have no open file.
+            if let Some(file) = extent.file.as_ref() {
+                // SAFETY: FFI call with a valid fd owned by the extent.
+                let res = unsafe { libc::fsync(file.as_raw_fd()) };
```
Comment (weltling) [inline block/src/formats/vmdk/engine_sync.rs]: Additional observation, here it should become `file.sync_all().map_err(AsyncIoError::Fsync)?;` instead of the unsafe block.

Thansk

File: block/src/formats/vmdk/mod.rs (hunk 2 above)
Comment (Sumynwa) [reply block/src/formats/vmdk/mod.rs]: Removed this unused field.

File: block/src/formats/vmdk/mod.rs (hunk 3 above)
Comment (Sumynwa) [reply block/src/formats/vmdk/mod.rs:48]: Done. I have used `query_device_size` on all extent file fds to make sure that the true host allocation is reported instead of the file's metadata len which can be sparse.

File: block/src/formats/vmdk/mod.rs (hunk 5 above)
Comment (Sumynwa) [reply block/src/formats/vmdk/mod.rs]: I get this now. I have fixed `physical_size` to use `query_device_size`.
Fixed & extended the tests to validate both sparse and actual block-allocated extents size.

File: block/src/formats/vmdk/flat.rs (hunk 4 above)
Comment (hsiangkao) [reply block/src/formats/vmdk/flat.rs]: Currently all high-level implementations (like kata) relies on absolute paths; so I hope absolute paths can be supported too.

Comment (Sumynwa) [reply block/src/formats/vmdk/flat.rs]: @hsiangkao  Yes absolute paths are supported right now. I have kept this similar to what Qemu allows, which is both relative and absolute.

Right now, I am not resolving any symlinks even in absolute path. That is something I will change.

File: block/src/formats/vmdk/engine_sync.rs (hunk 7 above)
Comment (Sumynwa) [reply block/src/formats/vmdk/engine_sync.rs]: Fixed.

File: block/src/formats/vmdk/flat.rs (hunk 6 above)
Comment (Sumynwa) [reply block/src/formats/vmdk/flat.rs:190]: I have kept both implementations and added test for both of them.

File: block/src/formats/vmdk/flat.rs (hunk 4 above)
Comment (Sumynwa) [reply block/src/formats/vmdk/flat.rs]: Updated to only refuse when the final component is a symlink.

**Comment by Sumynwa** [comment]
> Nice work. This is in good shape, and the extent path security handling is well done.
> 
> Besides the inline comments, the only gap I'd mention - no test for the new zero-sector/overflow rejection in FlatVmdk::new. Worth adding.
> 
> Thanks!

@weltling Added more unit tests for zero size/overflow in FlatVmdk::new

1. size_in_sectors == 0
2. size_in_sectors * 512 overflows u64
3. offset_in_sectors * 512 overflows u64
4. file_base_offset + length overflows u64
5. running sum of extent lengths overflows u64

File: block/src/formats/vmdk/engine_sync.rs [hunk 8]
```diff
@@ -0,0 +1,353 @@
...
+
+        let iovecs = op.iovecs();
+        let fd = file.as_raw_fd();
+        let file_offset = file_offset as libc::off_t;
+
+        // SAFETY: the iovec buffers are owned by `op` and remain valid for the
+        // duration of this call.
+        let res = unsafe {
+            if is_read {
+                libc::preadv(
```
Comment (weltling) [inline block/src/formats/vmdk/engine_sync.rs]: `main` now has `AlignedFile::read_vectored_at`/`write_vectored_at`. After rebasing, `single_extent_io` could just call them and drop `direct_io_is_aligned` plus the `unsafe` `preadv`/`pwritev` block. Also, misaligned `O_DIRECT` would no longer need `segment_io`, which could be kept for `spanning_io` only.

Thanks

**Review by weltling — CHANGES_REQUESTED** [review]
Nice work, thanks for addressing the batch. I left one inline note, lets do that before giving this a first nod.

Thanks

File: block/src/formats/vmdk/engine_sync.rs (hunk 8 above)
Comment (Sumynwa) [reply block/src/formats/vmdk/engine_sync.rs]: Makes sense. I will make the changes. Thanks.

Comment (Sumynwa) [reply block/src/formats/vmdk/engine_sync.rs]: Done. 
`single_extent_io` hands the read/write to AlignedFile's `read_vectored_at/write_vectored_at`.
`spanning_io` uses the `segment_io` as it is.

File: block/src/formats/vmdk/descriptor.rs [hunk 9]
```diff
@@ -0,0 +1,550 @@
...
+}
+
+// One extent line: `<access> <sectors> <type> "<file>" [offset]`.
+//
+// The optional trailing offset is the starting sector *within the backing
+// file*. It is non-zero when several extent lines reference the same file at
+// growing offsets (e.g. a >2GB file under `twoGbMaxExtentFlat`).
+//
+// Every field except `extent_type` is consumed by the flat extent layer added
+// in the following commit; the allow is removed there.
```
Comment (sboeuf) [inline block/src/formats/vmdk/descriptor.rs]: That's a general comment, but can you please reduce the amount of comments AI generated here. And make sure it doesn't introduce some weird `;` in the middle.

File: block/src/formats/vmdk/descriptor.rs [hunk 10]
```diff
@@ -0,0 +1,550 @@
...
+        VmdkDescriptorExtents,
+        VmdkDescriptorDdb,
+    )> {
+        let mut lines = input.lines();
+        let (header, last) = parse_header(&mut lines)?;
+        let (extents, ddb) = parse_extents_and_ddb(&mut lines, last)?;
+        Ok((header, extents, ddb))
+    }
+
+    // ---- parse_extents_and_ddb: valid inputs ----
```
Comment (sboeuf) [inline block/src/formats/vmdk/descriptor.rs]: We don't use that kind of separator in the project.

**Review by sboeuf — DISMISSED** [review]

File: block/src/formats/vmdk/descriptor.rs (hunk 10 above)
Comment (Sumynwa) [reply block/src/formats/vmdk/descriptor.rs]: @sboeuf i had tried to clean a lot of verbose text.
This was missed. Will trim down more. Thanks.

File: block/src/formats/vmdk/descriptor.rs (hunk 9 above)
Comment (Sumynwa) [reply block/src/formats/vmdk/descriptor.rs]: Fixed. Trimmed comments in all commits so every commit comment is clean and in expected format with no `;` in the middle.

File: block/src/formats/vmdk/descriptor.rs (hunk 10 above)
Comment (Sumynwa) [reply block/src/formats/vmdk/descriptor.rs]: Fixed.

**Comment by Sumynwa** [comment]
Thanks @weltling  @sboeuf  Addressed all review comments for now.

**Review by weltling — COMMENTED** [review]
`SyncCompletionQueue` landed in main. It wraps the completion `VecDeque` and  `EventFd`. On the next rebase `FlatVmdkSync` could adopt it, dropping its own `eventfd` and `completion_list`, so then it's consistent with the other sync engines.

Thanks

**Comment by Sumynwa** [comment]
> `SyncCompletionQueue` landed in main. It wraps the completion `VecDeque` and `EventFd`. On the next rebase `FlatVmdkSync` could adopt it, dropping its own `eventfd` and `completion_list`, so then it's consistent with the other sync engines.
> 
> Thanks

@weltling Changes have been done to use `CompletionCommon`.

File: block/src/formats/vmdk/mod.rs (hunk 3 above)
Comment (Sumynwa) [reply block/src/formats/vmdk/mod.rs:48]: @weltling  I have used `query_device_size` in the implementation of [physical_block_size](https://github.com/Sumynwa/cloud-hypervisor/blob/bcd929dfae1441fad5655064454b75c8262c4f56/block/src/formats/vmdk/flat.rs#L356)

File: block/src/formats/vmdk/flat.rs (hunk 6 above)
Comment (Sumynwa) [reply block/src/formats/vmdk/flat.rs:190]: Added tests for both `openat2` and `open_extent_walk` [here](https://github.com/Sumynwa/cloud-hypervisor/blob/bcd929dfae1441fad5655064454b75c8262c4f56/block/src/formats/vmdk/flat.rs#L477)

**Review by weltling — APPROVED** [review]
LGTM. Thanks for working through the reviews. No further technical concerns
so far. The shape is good. Thanks

**Review by hsiangkao — APPROVED** [review]
I'm not a cloud-hypervisor expert, but after a quick glance weeks ago I've seen.no major issue as well. Since this feature can bring more possibilities to various use cases (and especially erofs + kata use cases I'm concerned about), so I sincerely hope clh folks could consider this, thanks!

**Review by liuw — APPROVED** [review]

File: block/src/formats/vmdk/descriptor.rs [hunk 11]
```diff
@@ -0,0 +1,531 @@
...
+            .ok_or_else(|| {
+                io::Error::new(
+                    io::ErrorKind::InvalidInput,
+                    "Cannot retrieve parent directory of the file",
+                )
+            })?
+            .to_string_lossy()
+            .to_string();
+
+        // A valid descriptor is at least a few bytes of text.
```
Comment (russell-islam) [inline block/src/formats/vmdk/descriptor.rs]: Is it few bytes or at least 4 bytes, Plz be clear on the comment.

File: block/src/formats/vmdk/descriptor.rs [hunk 12]
```diff
@@ -0,0 +1,531 @@
...
+    String::from_utf8(buf).map_err(|_| {
+        io::Error::new(
+            io::ErrorKind::InvalidData,
+            "VMDK descriptor is not valid UTF-8",
+        )
+    })
+}
+
+pub(crate) fn parse_header<'a>(
+    lines: &mut Lines<'a>,
```
Comment (russell-islam) [inline block/src/formats/vmdk/descriptor.rs:127]: Why do you need a mut reference, Do you make any changes?

File: block/src/formats/vmdk/descriptor.rs (hunk 11 above)
Comment (Sumynwa) [reply block/src/formats/vmdk/descriptor.rs]: Updated. Thanks.

File: block/src/formats/vmdk/descriptor.rs (hunk 12 above)
Comment (Sumynwa) [reply block/src/formats/vmdk/descriptor.rs:127]: In both `parse_header` and `parse_extents`, I am passing Lines iterator from String::lines().
Need the mut here for calling iterators next() method.

**Comment by liuw** [comment]
@Sumynwa

Some of the tests are only run during merge. See this https://github.com/cloud-hypervisor/cloud-hypervisor/actions/runs/30830817138/job/91746211603

It looks like a genuine issue.

**Comment by Sumynwa** [comment]
> @Sumynwa
> 
> 
> 
> Some of the tests are only run during merge. See this https://github.com/cloud-hypervisor/cloud-hypervisor/actions/runs/30830817138/job/91746211603
> 
> 
> 
> It looks like a genuine issue.

Sure, I will take a look at it.

**Comment by Sumynwa** [comment]
The [integration-arm64](https://github.com/cloud-hypervisor/cloud-hypervisor/actions/runs/30830817138/job/91746211603#step:6:481) test is failing.

Failure: `test_virtio_block_vmdk_two_gb_max_extent_flat_spanning_direct`
Error:
`4: I/O error (path=/tmp/chypjN0l/spanning.vmdk op=open)`
`5: Invalid argument (os error 22)`
 
The problem is I am using `O_DIRECT` flag in this test. The VMDK descriptor text file is being kept at `guest.tmpDir` which is the `tmpfs` inside the docker. From what I understood, we will face this issue when `tmpfs does not support O_DIRECT` which looks like the case in the Arm64 CI Runner.
One other test I added `test_virtio_block_direct_io_data_disk_4k_vmdk` is [passing](https://github.com/cloud-hypervisor/cloud-hypervisor/actions/runs/30830817138/job/91746211603#step:6:481) on this runner. In this test, the files ARE NOT on tmpfs.
 
A similar comment is mentioned in integration test [test_virtio_block_direct_and_firmware](https://github.com/cloud-hypervisor/cloud-hypervisor/blob/dcea656a722cab1b24c1d7c48fa2b12a7276f04c/cloud-hypervisor/tests/integration.rs#L2051)

**Comment by liuw** [comment]
> The [integration-arm64](https://github.com/cloud-hypervisor/cloud-hypervisor/actions/runs/30830817138/job/91746211603#step:6:481) test is failing.
> 
> Failure: `test_virtio_block_vmdk_two_gb_max_extent_flat_spanning_direct` Error: `4: I/O error (path=/tmp/chypjN0l/spanning.vmdk op=open)` `5: Invalid argument (os error 22)`
> 
> The problem is I am using `O_DIRECT` flag in this test. The VMDK descriptor text file is being kept at `guest.tmpDir` which is the `tmpfs` inside the docker. From what I understood, we will face this issue when `tmpfs does not support O_DIRECT` which looks like the case in the Arm64 CI Runner. One other test I added `test_virtio_block_direct_io_data_disk_4k_vmdk` is [passing](https://github.com/cloud-hypervisor/cloud-hypervisor/actions/runs/30830817138/job/91746211603#step:6:481) on this runner. In this test, the files ARE NOT on tmpfs.
> 
> A similar comment is mentioned in integration test [test_virtio_block_direct_and_firmware](https://github.com/cloud-hypervisor/cloud-hypervisor/blob/dcea656a722cab1b24c1d7c48fa2b12a7276f04c/cloud-hypervisor/tests/integration.rs#L2051)

I guess we can remove O_DIRECT in the test case then?

**Comment by Sumynwa** [comment]
@liuw  I am using `~/workloads` for the artifact directory when using `O_DIRECT` (similar to [this](https://github.com/cloud-hypervisor/cloud-hypervisor/blob/dcea656a722cab1b24c1d7c48fa2b12a7276f04c/cloud-hypervisor/tests/integration.rs#L2054))

File: cloud-hypervisor/tests/common/tests_wrappers.rs [hunk 13]
```diff
@@ -2170,6 +2170,303 @@ pub fn _test_virtio_block_dynamic_vhdx_expand(guest: &Guest) {
...
+    let output = cloud_child.wait_with_output().unwrap();
+
+    handle_child_output(r, &output);
+
+    disk_check_consistency(vmdk_path_str, None);
+}
+
+pub(crate) fn _test_virtio_block_vmdk_extent_spanning(guest: &Guest, direct: bool) {
+    // 3 GiB virtual size => a 2 GiB first extent + a 1 GiB second extent.
+    // `direct=on` opens the extents with `O_DIRECT`, which tmpfs rejects with
```
Comment (Sumynwa) [inline cloud-hypervisor/tests/common/tests_wrappers.rs:2343]: @liuw Please refer to this. Thanks.

Comment (liuw) [reply cloud-hypervisor/tests/common/tests_wrappers.rs:2343]: Okay, this makes sense.

File: block/src/lib.rs [hunk 14]
```diff
@@ -538,10 +541,23 @@ pub fn open_disk_image(path: &Path, options: &OpenOptions) -> BlockResult<File>
...
+            Ok(0) => break,
+            Ok(n) => filled += n,
+            Err(ref e) if e.kind() == io::ErrorKind::Interrupted => continue,
+            Err(e) => {
+                return Err(
+                    BlockError::new(BlockErrorKind::Io, e).with_op(ErrorOp::DetectImageType)
+                );
+            }
+        }
+    }
```
Comment (rbradford) [inline block/src/lib.rs:560]: Image type detection is deprecated. This should not have been merged with this in.

File: docs/landlock.md [hunk 15]
```diff
@@ -62,6 +62,31 @@ Landlock can also be enabled during `vm.create` request by passing a config like
...
+### Multi-file disk formats (VMDK)
+
+For most block backends (raw, qcow2, VHD) including existing VMDK, landlock
+grants access to the `--disk path=` path value.
+
+For a Flat VMDK, its `path=` points at a small text descriptor whose
+data lives in one or more separate extent files. Granting only the descriptor
+currently leaves those extents unreachable under landlock.
+
+The process launching Cloud-Hypervisor must grant these extent paths explicitly
```
Comment (rbradford) [inline docs/landlock.md:74]: We never hyphenate and capitalise like that.

**Review by rbradford — COMMENTED** [review]
I didn't review in detail as this was already merged but I found some things that should have been caught on the review.

File: docs/landlock.md (hunk 15 above)
Comment (Sumynwa) [reply docs/landlock.md:74]: I will fix this.

File: block/src/lib.rs (hunk 14 above)
Comment (Sumynwa) [reply block/src/lib.rs:560]: I will check this again and update.
The block backend did went under a lot of api changes when i was working on this and this probably is the reason it went in.
Thanks.

Comment (Sumynwa) [reply block/src/lib.rs:560]: @rbradford I will need help here to understand more since I am new to the codebase.
I followed the code path:
[make_virtio_block_device](https://github.com/cloud-hypervisor/cloud-hypervisor/blob/a1fcb9f790616ac615f66de73be540b0b20844b1/vmm/src/device_manager.rs#L2690) => [open_disk](https://github.com/cloud-hypervisor/cloud-hypervisor/blob/a1fcb9f790616ac615f66de73be540b0b20844b1/block/src/factory.rs#L86) => [detect_image_type](https://github.com/cloud-hypervisor/cloud-hypervisor/blob/a1fcb9f790616ac615f66de73be540b0b20844b1/block/src/factory.rs#L95)

The [detect_image_type](https://github.com/cloud-hypervisor/cloud-hypervisor/blob/a1fcb9f790616ac615f66de73be540b0b20844b1/block/src/lib.rs#L542) implementation used `AlignedFile::read_exact_at` with O_DIRECT.
The problem I faced with VMDK (file opened here being the descriptor text file size 512 bytes) was:
```
    cloud-hypervisor:   0.004684s: <main> ERROR:/cloud-hypervisor/cloud-hypervisor/src/lib.rs:24 -- Fatal error: VmBoot(VmBoot(DeviceManager(Disk(BlockError { kind: Io, source: Some(Error { kind: UnexpectedEof, message: "failed to fill whole buffer" }), ctx: Some(ErrorContext { path: None, offset: None, op: Some(DetectImageType) }) }))))
    Error: Cloud Hypervisor exited with the following chain of errors:
      0: Error booting VM
      1: The VM could not boot
      2: Error from device manager
      3: Cannot open disk path
      4: I/O error (op=detect_image_type)
      5: failed to fill whole buffer
```
Hence this change was made.

Comment (Sumynwa) [reply block/src/lib.rs:560]: I have made a fix that almost removes all changes done here except `is_flat_vmdk` to detect VMDK image type.
But it is needed here otherwise the format will error out with "failed to fill whole buffer" for small descriptor texts.

Comment (liuw) [reply block/src/lib.rs:560]: What Rob means that is that machinery is going away. Users should provide the type while they add the disk.

What's left for CH to do is to validate the signature (+whatever else required) of the disk image against what the user says.

Comment (Sumynwa) [reply block/src/lib.rs:560]: Yes makes sense 
But with existing code for vmdk to work i still need to patch detect_image_type.

Comment (rbradford) [reply block/src/lib.rs:560]: I have a PR that I will be publishing to delete `detect_image_type`.
